lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs

```diff
@@ -3070,80 +3064,36 @@ balanceTransaction
     (ApiT wid)
     body = do
     -- NOTE: Ideally we'd read @pp@ and @era@ atomically.
-    pp <- liftIO $ NW.currentProtocolParameters nl
-    era <- liftIO $ NW.currentNodeEra nl
+    era <- liftIO $ NW.currentNodeEra netLayer
+    pp <- liftIO (currentProtocolParameters netLayer)
+    Write.AnyRecentEra recentEra <- guardIsRecentEra era
+
     withWorkerCtx ctx wid liftE liftE $ \wrk -> do
-        (walletUTxO, wallet, _txs) <- handler $ W.readWalletUTxO @_ @s wrk
+        (utxo, wallet, _txs) <- handler $ W.readWalletUTxO wrk
         timeTranslation <- liftIO $ toTimeTranslation (timeInterpreter netLayer)
-        let mkPartialTx
-                :: forall era. Write.IsRecentEra era => Cardano.Tx era
-                -> Handler (Write.PartialTx era)
-            mkPartialTx tx = do
-                utxo <- fmap Write.toCardanoUTxO $ mkLedgerUTxO $ body ^. #inputs
-                pure $ Write.PartialTx
-                    tx
-                    utxo
-                    (fromApiRedeemer <$> body ^. #redeemers)
-              where
-                -- NOTE: There are a couple of spread-out pieces of logic
-                -- dealing with the choice of era, most prominantly: tx, utxo,
-                -- pparams / current node era. It /might/ be neater to have a
-                -- single function dedicated to this choice instead; something
-                -- like
-                -- @@
-                --      chooseEra
-                --          :: InRecentEra Tx
-                --          -> InRecentEra UTxO
-                --          -> InRecentEra PParams
-                --          -> (IsRecentEra era
-                --              => Tx era
-                --              -> UTxO era
-                --              -> PParams era
-                --              -> res)
-                --          -> res
-                -- @@
-
-                mkRecentEra :: Handler (Write.RecentEra era)
-                mkRecentEra = case Cardano.cardanoEra @era of
-                    Cardano.ConwayEra -> pure Write.RecentEraConway
-                    Cardano.BabbageEra -> pure Write.RecentEraBabbage
-                    _ -> liftHandler $ throwE $ Write.ErrOldEraNotSupported era
-
-                mkLedgerUTxO
-                    :: [ApiExternalInput n]
-                    -> Handler (Write.UTxO (ShelleyLedgerEra era))
-                mkLedgerUTxO ins = do
-                    recentEra <- mkRecentEra
-                    pure
-                        . Write.utxoFromTxOutsInRecentEra recentEra
-                        . map fromExternalInput
-                        $ ins
+
+        partialTx <- parsePartialTx recentEra
 
         let balanceTx
                 :: forall era. Write.IsRecentEra era
                 => Write.PartialTx era
                 -> Handler (Cardano.Tx era)
-            balanceTx partialTx =
+            balanceTx ptx =
```

Comment:
IMO `ptx` is less readable than `partialTx`

Reply:
Ah - I went with `ptx` to avoid shadowing.

Arguably I don't need this local `let` function:
```haskell
        let balanceTx
                :: forall era. Write.IsRecentEra era
                => Write.PartialTx era
                -> Handler (Cardano.Tx era)
            balanceTx partialTx =
                liftHandler $ fst <$> Write.balanceTransaction @_ @IO @s
                    (MsgWallet . W.MsgBalanceTx >$< wrk ^. W.logger)
                    (Write.UTxOAssumptions
                        genInpScripts
                        mScriptTemplate
                        (txWitnessTagFor @k)
                        )
                    (Write.unsafeFromWalletProtocolParameters @era pp)
                    timeTranslation
                    (Write.constructUTxOIndex utxo)
                    (W.defaultChangeAddressGen argGenChange)
                    (getState wallet)
                    partialTx
```
and if I inlined it the shadowing problem would disappear, but I thought it kind-of made sense conceptually 🤔

Reply:
Now inlined.